### ISSUE TITLE
Internal DNS not available on rancher 2.0 

### ISSUE THREAD

**Issue by Rgils** [rancher/rancher#13504]
**Rancher versions:**
rancher/server or rancher/rancher: 2.0
rancher/agent or rancher/rancher-agent:
```
:~$ docker ps
CONTAINER ID        IMAGE                                             COMMAND                  CREATED              STATUS              PORTS                                      NAMES
c2c759918876        c2ce1ffb51ed                                      "/dnsmasq-nanny -v=2…"   About a minute ago   Up About a minute                                              k8s_dnsmasq_kube-dns-7dfdc4897f-bv9rp_kube-system_77b78a88-5243-11e8-930c-000c2969ab2a_2240
9219c15917fe        6f7f2dc7fab5                                      "/sidecar --v=2 --lo…"   6 minutes ago        Up 6 minutes                                                   k8s_sidecar_kube-dns-7dfdc4897f-bv9rp_kube-system_77b78a88-5243-11e8-930c-000c2969ab2a_1759
a3c2dd8355b8        80cc5ea4b547                                      "/kube-dns --domain=…"   7 minutes ago        Up 7 minutes                                                   k8s_kubedns_kube-dns-7dfdc4897f-bv9rp_kube-system_77b78a88-5243-11e8-930c-000c2969ab2a_965
c1ad55ecc0a1        centos                                            "/bin/bash"              5 days ago           Up 5 days                                                      k8s_centos_centos-7954dd9476-b8p54_all_8a7a4bf1-53ca-11e8-a79a-000c2969ab2a_0
232bee34f818        rancher/pause-amd64:3.1                           "/pause"                 5 days ago           Up 5 days                                                      k8s_POD_centos-7954dd9476-b8p54_all_8a7a4bf1-53ca-11e8-a79a-000c2969ab2a_0
2c6443b236ff        8cfec7659f1d                                      "run.sh"                 7 days ago           Up 7 days                                                      k8s_cluster-register_cattle-cluster-agent-6cdd9c5776-4s9zq_cattle-system_804971ad-52c1-11e8-a79a-000c2969ab2a_0
625fcd6403f1        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_cattle-cluster-agent-6cdd9c5776-4s9zq_cattle-system_804971ad-52c1-11e8-a79a-000c2969ab2a_0
687a36f8d19a        rancher/hyperkube:v1.10.1-rancher2                "/opt/rke/entrypoint…"   7 days ago           Up 7 days                                                      kube-apiserver
d81e0b95664a        rancher/coreos-etcd:v3.1.12                       "/usr/local/bin/etcd…"   7 days ago           Up 7 days                                                      etcd
a857efa30733        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_kube-dns-7dfdc4897f-bv9rp_kube-system_77b78a88-5243-11e8-930c-000c2969ab2a_0
5fa34620d163        rancher/nginx-ingress-controller-defaultbackend   "/server"                7 days ago           Up 7 days                                                      k8s_default-http-backend_default-http-backend-564b9b6c5b-xjrt2_ingress-nginx_7ac01a79-5243-11e8-930c-000c2969ab2a_0
5a274f06f0be        rancher/cluster-proportional-autoscaler-amd64     "/cluster-proportion…"   7 days ago           Up 7 days                                                      k8s_autoscaler_kube-dns-autoscaler-6c4b786f5-rrmsz_kube-system_7848e989-5243-11e8-930c-000c2969ab2a_0
62f224369b04        rancher/coreos-flannel                            "/opt/bin/flanneld -…"   7 days ago           Up 7 days                                                      k8s_kube-flannel_canal-bh7j5_kube-system_7612c014-5243-11e8-930c-000c2969ab2a_0
b56a3d0ed6f1        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_default-http-backend-564b9b6c5b-xjrt2_ingress-nginx_7ac01a79-5243-11e8-930c-000c2969ab2a_0
9d6e249a4dfc        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_kube-dns-autoscaler-6c4b786f5-rrmsz_kube-system_7848e989-5243-11e8-930c-000c2969ab2a_0
333381c7eada        rancher/calico-cni                                "/install-cni.sh"        7 days ago           Up 7 days                                                      k8s_install-cni_canal-bh7j5_kube-system_7612c014-5243-11e8-930c-000c2969ab2a_0
d406a1e60e7b        rancher/calico-node                               "start_runit"            7 days ago           Up 7 days                                                      k8s_calico-node_canal-bh7j5_kube-system_7612c014-5243-11e8-930c-000c2969ab2a_0
5c2216484d45        8cfec7659f1d                                      "run.sh"                 7 days ago           Up 7 days                                                      k8s_agent_cattle-node-agent-tjdpc_cattle-system_7d8b311e-5243-11e8-930c-000c2969ab2a_0
71481f1fac31        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_cattle-node-agent-tjdpc_cattle-system_7d8b311e-5243-11e8-930c-000c2969ab2a_0
885e68fed3ad        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_nginx-ingress-controller-htvw9_ingress-nginx_7abd433a-5243-11e8-930c-000c2969ab2a_0
0393d0918657        rancher/pause-amd64:3.1                           "/pause"                 7 days ago           Up 7 days                                                      k8s_POD_canal-bh7j5_kube-system_7612c014-5243-11e8-930c-000c2969ab2a_0
2c015c79ba7e        rancher/hyperkube:v1.10.1-rancher2                "/opt/rke/entrypoint…"   7 days ago           Up 7 days                                                      kube-proxy
7ea6ed297229        rancher/hyperkube:v1.10.1-rancher2                "/opt/rke/entrypoint…"   7 days ago           Up 7 days                                                      kubelet
56d641dea902        rancher/hyperkube:v1.10.1-rancher2                "/opt/rke/entrypoint…"   7 days ago           Up 7 days                                                      kube-scheduler
2b1a80503786        rancher/hyperkube:v1.10.1-rancher2                "/opt/rke/entrypoint…"   7 days ago           Up 7 days                                                      kube-controller-manager
eb717ee0a551        rancher/rancher                                   "rancher --http-list…"   7 days ago           Up 7 days           0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   wizardly_leakey

```
```
 docker logs -f c2c759918876
I0515 13:20:05.288223       1 main.go:76] opts: {{/usr/sbin/dnsmasq [-k --cache-size=1000 --log-facility=- --server=/cluster.local/127.0.0.1#10053 --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053] true} /etc/k8s/dns/dnsmasq-nanny 10000000000}
I0515 13:20:05.288378       1 nanny.go:94] Starting dnsmasq [-k --cache-size=1000 --log-facility=- --server=/cluster.local/127.0.0.1#10053 --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053]
I0515 13:20:05.574666       1 nanny.go:119]
W0515 13:20:05.574699       1 nanny.go:120] Got EOF from stdout
I0515 13:20:05.574720       1 nanny.go:116] dnsmasq[19]: started, version 2.78 cachesize 1000
I0515 13:20:05.574739       1 nanny.go:116] dnsmasq[19]: compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-DNSSEC loop-detect inotify
I0515 13:20:05.574751       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.1#10053 for domain ip6.arpa
I0515 13:20:05.574755       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa
I0515 13:20:05.574758       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.1#10053 for domain cluster.local
I0515 13:20:05.574764       1 nanny.go:116] dnsmasq[19]: reading /etc/resolv.conf
I0515 13:20:05.574768       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.1#10053 for domain ip6.arpa
I0515 13:20:05.574771       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa
I0515 13:20:05.574775       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.1#10053 for domain cluster.local
I0515 13:20:05.574779       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.53#53
I0515 13:20:05.574795       1 nanny.go:116] dnsmasq[19]: read /etc/hosts - 7 addresses
I0515 13:20:25.255275       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:20:35.270254       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:20:45.282302       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:20:55.298429       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:21:05.311510       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:21:15.326011       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:21:25.334450       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:21:35.349728       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:21:45.356621       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:21:55.365439       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
I0515 13:22:05.371858       1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)
```
```
~$ cat /etc/netplan/50-cloud-init.yaml
# This file is generated from information provided by
# the datasource.  Changes to it will not persist across an instance.
# To disable cloud-init's network configuration capabilities, write a file
# /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg with the following:
# network: {config: disabled}
network:
    ethernets:
        ens160:
            addresses:
            - 172.28.250.11/22
            gateway4: 172.28.248.1
            nameservers:
                addresses:
                - 172.26.250.1
                - 172.26.250.2
                search:
                - internal
            optional: true
    version: 2
```
```
~$ cat /etc/docker/daemon.json
{
    "dns": ["172.26.250.1","172.26.250.2"]
}
```
```
~$ cat /etc/environment
PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games"
http_proxy="http://proxy.internal:3128/"
https_proxy="http://proxy.internal:3128/"
ftp_proxy="http://proxy.internal:3128/"
no_proxy="localhost,127.0.0.1,172.28.250.11,172.28.250.14,.internal"
HTTP_PROXY="http://proxy.internal:3128/"
HTTPS_PROXY="http://proxy.internal:3128/"
FTP_PROXY="http://proxy.internal:3128/"
```
```
:~$ cat /etc/systemd/system/docker.service.d/http-proxy.conf
[Service]
Environment="HTTP_PROXY=http://proxy.internal:3128/" "NO_PROXY=localhost,127.0.0.1,0.0.0.0"

```
**Docker version: (`docker version`,`docker info` preferred)**
```
~$ docker info
Containers: 41
 Running: 25
 Paused: 0
 Stopped: 16
Images: 18
Server Version: 17.12.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 9f9c96235cc97674e935002fc3d78361b696a69e
init version: v0.13.0 (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-20-generic
Operating System: Ubuntu 18.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.789GiB
Name: luke1
ID: L77A:SI7V:T33N:4WHM:DLFN:GXHX:7IVT:LKPT:D5MB:2KZW:4WM4:7CEA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
HTTP Proxy: http://proxy.internal:3128/
No Proxy: localhost,127.0.0.1,0.0.0.0
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```
**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
```
~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
Bare-metal
**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node and multinode

**Steps to Reproduce:**
Install Ubuntu 18.04 (static network config, with internal dns), configure Proxy on /etc/systemd/system/docker.service.d/http-proxy.conf, /etc/environment and dns in /etc/docker/daemon.json, reload system and run the instalation of rancher server, create custom cluster and start multiple pods, alias doesn't works as ping 10.
**Results:**
Not dns resolving inside the cluster:
instance centos:7
```
[root@centos-5b54f6f9f4-pv9x5 /]# cat /etc/resolv.conf
nameserver 10.43.0.10
search default.svc.cluster.local svc.cluster.local cluster.local internal
options ndots:5
[root@centos-5b54f6f9f4-pv9x5 /]# ping 10.42.0.10
PING 10.42.0.10 (10.42.0.10) 56(84) bytes of data.
^C
--- 10.42.0.10 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2049ms
```

**Comment by vincent99** [comment]
Cluster IPs don't respond to pings.

> I0515 13:20:05.574779       1 nanny.go:116] dnsmasq[19]: using nameserver 127.0.0.53#53
> I0515 13:20:05.574795      1 nanny.go:116] dnsmasq[19]: read /etc/hosts - 7 addresses
> I0515 13:20:25.255275      1 nanny.go:116] dnsmasq[19]: Maximum number of concurrent DNS queries reached (max: 150)

Suggests that the DNS container picked up 127.0.0.1 as the recurser and is immediately calling itself recursively.

**Comment by Rgils** [comment]
ping to other IPs (inside and outside)
```
[root@centos-5b54f6f9f4-pv9x5 /]# ping  10.42.0.13
PING 10.42.0.13 (10.42.0.13) 56(84) bytes of data.
64 bytes from 10.42.0.13: icmp_seq=1 ttl=64 time=0.023 ms
64 bytes from 10.42.0.13: icmp_seq=2 ttl=64 time=0.038 ms
^C
--- 10.42.0.13 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1022ms
rtt min/avg/max/mdev = 0.023/0.030/0.038/0.009 ms
[root@centos-5b54f6f9f4-pv9x5 /]# ping 10.42.0.12
PING 10.42.0.12 (10.42.0.12) 56(84) bytes of data.
64 bytes from 10.42.0.12: icmp_seq=1 ttl=63 time=0.066 ms
64 bytes from 10.42.0.12: icmp_seq=2 ttl=63 time=0.066 ms
^C
[root@centos-5b54f6f9f4-pv9x5 /]# ping 172.28.250.10
PING 172.28.250.10 (172.28.250.10) 56(84) bytes of data.
64 bytes from 172.28.250.10: icmp_seq=1 ttl=63 time=0.261 ms
64 bytes from 172.28.250.10: icmp_seq=2 ttl=63 time=0.230 ms
^C
--- 172.28.250.10 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1026ms
rtt min/avg/max/mdev = 0.230/0.245/0.261/0.022 ms
```
It's just the DNS service miss-configuration somehow who fall down 

Probably related: #13306 but pointing to google dns (or ours Internal DNS) will just solve external IPs not internal services on the cluster

**Comment by iwilltry42** [comment]
I also had problems with Kube-DNS not working in a fresh cluster... Here's what I found out:
Kube-DNS/Core-DNS have problems with systemd-resolved based systems (like Ubuntu 18.04).
See this issue: https://github.com/kubernetes/kubeadm/issues/273
Workaround in this issue: https://github.com/kubernetes/kubeadm/issues/787
On the hosts workaround via: 
`sudo rm /etc/resolv.conf
sudo ln -s /run/systemd/resolve/resolv.conf /etc/resolv.conf`

**Comment by superseb** [comment]
Tracked in https://github.com/rancher/rancher/issues/13782 and https://github.com/rancher/rancher/issues/13888
